### PR TITLE
Render YamlRawValue block scalars in the lambda textarea (fixes #428)

### LINKED ISSUE #428: [Bug] Lambda field renders as "[object Object]" when YAML uses block scalar (|-)
## What

When a `binary_sensor: - platform: template` entry has its `lambda:` field defined with a YAML block scalar (`|-` / `|`), the visual editor renders the input as the literal text `[object Object]` instead of the lambda's source code.

```yaml
binary_sensor:
  - platform: template
    name: "${upper_devicename} Opening"
    id: ${id_prefix}_opening_sensor
    lambda: |-
      return id(${id_prefix}_moving) && id(${id_prefix}_opening) && !id(${id_prefix}_opened).state && !id(${id_prefix}_closed).state ? true : false;
```

The Lambda input shows `[object Object]` (see screenshot). The YAML pane on the right preserves the source unchanged — round-trip from disk → editor → disk would replace the lambda body with the literal string `[object Object]` if the user clicks Save without noticing.

## Reproduction

1. Open a device whose YAML has a `binary_sensor: - platform: template` with a `lambda: |-` block scalar.
2. Click the section to open the visual editor for that entry.
3. The Lambda field shows `[object Object]`.

## Expected

The Lambda input shows the lambda's source code (the contents of the `|-` block) as a multi-line string, identical to what the user typed in YAML.

## Likely root cause

A YAML block scalar parser somewhere upstream of the form input is producing a wrapper object — likely a `{ value: "...", style: "block" }` or similar shape from the YAML AST — and the form is calling `String(value)` on it (or an equivalent implicit string conversion) which yields `[object Object]`.

Two places to check:

1. **Section-config values pipeline** (`device-section-config.ts` → `_values`) — where the YAML is parsed into the form's values dict. If the parser keeps block-scalar metadata, the leaf value isn't a plain string.
2. **The lambda renderer** (`config-entry-renderers.ts`, the renderer for `type: "lambda"`) — should defensively coerce via `extractScalarString(value)` (or whatever helper handles block scalars) instead of relying on the parser to flatten.

The right fix is probably defensive at the renderer: any string-typed entry should accept `string | { value: string }` and unwrap. But the fundamental story is that the parser is leaking a non-leaf shape downstream.

## Out of scope

- Templatable string fields (`{ '!lambda': "..." }`) — those are the YAML tag form, a different shape, and the form already handles them via the templatable-input branch.
- Other multi-line YAML fields (`packages:` includes, `script:` action lists) — same root cause likely affects them; this issue tracks the lambda case as the most visible example. Other fields can be folded into the same fix.

## Workaround

Switch the lambda to single-line form (`lambda: "return ...;"`) — the inline string is a plain scalar and renders correctly. Painful for non-trivial lambdas.

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

Fixes esphome/device-builder#428.

The form's textarea renderer (lambda / json fields) was calling `String(ctx.getAt(path) ?? "")` directly on the value dict. When the parser captured a YAML block scalar (`lambda: |-` ... body) into a `YamlRawValue` so the on-disk `|-` style would round-trip, `String()` produced the default `"[object Object]"` — what the user saw in the screenshot on the issue. A no-op save would then have replaced the lambda body with the literal text `"[object Object]"` and corrupted the YAML.

## Fix

Two parts:

1. **`YamlRawValue` exposes editor-friendly accessors.** New `body` / `indent` / `toString` getters so `String(rawValue)` yields the dedented body string. The indent algorithm walks every non-blank line shrinking the common prefix; blank lines are ignored (so a blank middle line doesn't collapse the indent to empty); the body strips that common prefix verbatim, preserving any deeper indents inside the user's lambda. This protects every `String(value)` site downstream, not just the textarea.

2. **`renderTextareaField` re-wraps on edit.** The renderer detects the `YamlRawValue` and on `@input` calls a new `YamlRawValue.fromBodyText(text, original)` helper that re-applies the original common indent and preserves the inline `|-` header. On save the serializer pastes the re-indented lines back in their original block-scalar shape — a no-op open + save round-trip is byte-identical.

The fix is local to two files:
- `src/util/yaml-serialize.ts` — `body` / `indent` / `toString` / `fromBodyText` on `YamlRawValue`.
- `src/components/device/config-entry-renderers.ts` — `renderTextareaField` uses the wrapper.

## Tests

- 16 unit tests in `test/util/yaml-raw-value.test.ts` — `toString` / template-literal path, `indent` algorithm (common-prefix shrinking, blank-line ignoring, no-shared-indent fallback, single line), `body` dedenting (deeper-indent preservation, blank-line preservation), `fromBodyText` (re-apply indent, round-trip identity, blank lines emit empty without trailing whitespace, zero-prefix edge case, list-rooted-block undefined-header preservation).
- 2 integration tests in `test/util/yaml-section-values.test.ts` pinning the issue #428 fix end-to-end:
  - lambda block scalar re-saved without editing is byte-identical (no drift)
  - lambda edited via the textarea path comes back as a properly-indented `|-` block

All 857 existing FE tests pass.

**Related issue or feature (if applicable):**

- fixes esphome/device-builder#428

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue) — `bugfix`
- [ ] New feature (non-breaking change which adds functionality) — `new-feature`
- [ ] Enhancement to an existing feature — `enhancement`
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — `breaking-change`
- [ ] Refactor (no behaviour change) — `refactor`
- [ ] Documentation only — `docs`
- [ ] Maintenance / chore — `maintenance`
- [ ] CI / workflow change — `ci`
- [ ] Dependencies bump — `dependencies`

## Checklist

- [x] The code change is tested and works locally.
- [x] `npm run lint` passes.
- [x] `npm run test` passes.
- [x] Tests have been added to verify that the new code works (where applicable).

### CODE REVIEW

File: src/components/device/config-entry-renderers.ts [hunk 1]
```diff
@@ -449,8 +458,13 @@ export function renderTextareaField(
...
         .value=${value}
         placeholder=${String(entry.default_value ?? "")}
-        @input=${(e: Event) =>
-          ctx.emitChange(path, (e.target as HTMLTextAreaElement).value)}
+        @input=${(e: Event) => {
+          const text = (e.target as HTMLTextAreaElement).value;
+          ctx.emitChange(
+            path,
+            isRaw ? YamlRawValue.fromBodyText(text, raw) : text,
+          );
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/components/device/config-entry-renderers.ts:466]: False positive — TypeScript 5.9 + `strict` mode handles this correctly via "Aliased Conditions and Discriminants" (TS 4.4+). The `const isRaw = raw instanceof YamlRawValue` narrowing propagates into both ternary branches, so `raw.body` and `fromBodyText(text, raw)` see `raw` narrowed to `YamlRawValue`. `npm run lint` (tsc --noEmit, strict) passes cleanly. No fix needed.

File: src/util/yaml-serialize.ts [hunk 2]
```diff
@@ -48,6 +48,82 @@ export class YamlRawValue {
...
+    return common;
+  }
+
+  /**
+   * The block-scalar body as the user typed it semantically,
+   * with the common indent stripped. ``YamlRawValue("    return foo;\n
+   * return bar;")`` displays as ``"return foo;\nreturn bar;"``
+   * — which is what a textarea / lambda editor wants to render.
+   *
+   * Round-trip pairing: ``new YamlRawValue.fromBodyText(body, original)``
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/util/yaml-serialize.ts]: Good catch — fixed in b313fc8. The example was syntactically wrong: `YamlRawValue("...")` was missing `new`, and `new YamlRawValue.fromBodyText(...)` shouldn't have `new` since it's a static factory. Rephrased the docstring to describe the `lines` array shape directly so the example doesn't depend on a syntactically-correct constructor call.

File: test/util/yaml-section-values.test.ts [hunk 3]
```diff
@@ -722,6 +723,79 @@ describe("parseYamlSectionValues / updateSectionInYaml — block scalars and com
...
+    const after = updateSectionInYaml(
+      yaml,
+      "binary_sensor.template",
+      values,
+      2,
+    );
+    expect(after).toContain("lambda: |-");
+    expect(after).toContain(
+      "return id(moving) && id(opening) && !id(opened).state ? true : false;",
+    );
```
[the remark this answers is not in the document]
Comment (bdraco) [reply test/util/yaml-section-values.test.ts]: Good catch — fixed in b313fc8. Upgraded to byte-equality (`expect(after).toBe(yaml)`) plus a belt-and-suspenders `toContain` on the lambda block itself.

Hit an interesting wrinkle while doing it: byte-equality on the whole YAML revealed a pre-existing serializer behaviour where `name: "Driveway Opening"` loses its quotes on round-trip (the scalar serializer drops unneeded quotes for scalars that don't strictly need them). Unrelated to #428's lambda issue, so the test fixture now uses bare `name: Driveway Opening` and the byte-equality assertion is meaningful for what this PR is actually testing — the lambda block's preservation. The `toContain` on the lambda block itself ensures even if a future serializer change reformats surrounding keys, the lambda still has to survive byte-identical.
